File: iMEGA/Camera uploads/UploadOperations/CameraUploadNodeLoadOperation.m

#import "CameraUploadNodeLoadOperation.h"
#import "CameraUploadRequestDelegate.h"
#import "MEGAError+MNZCategory.h"

static NSString * const CameraUploadsNodeHandleKey = @"CameraUploadsNodeHandle";

@interface CameraUploadNodeLoadOperation ()

@property (copy, nonatomic) CameraUploadNodeLoadCompletionHandler completion;

@end

@implementation CameraUploadNodeLoadOperation

- (instancetype)initWithLoadCompletion:(CameraUploadNodeLoadCompletionHandler)completion {
    self = [super init];
    if (self) {
        _completion = completion;
    }
    
    return self;
}

- (void)start {
    if (self.cancelled) {
        if (self.completion) {
            self.completion(nil, nil);
        }
        
        [self finishOperation];
        return;
    }
    
    [self startExecuting];
    
    MEGANode *node = [self restoreExistingCameraUploadNode];
    if (node) {
        MEGALogDebug(@"[Camera Upload] existing camera upload node is restored");
        if (self.completion) {
            self.completion(node, nil);
        }
        
        [self finishOperation];
    } else {
        CameraUploadRequestDelegate *delegate = [[CameraUploadRequestDelegate alloc] initWithCompletion:^(MEGARequest * _Nonnull request, MEGAError * _Nonnull error) {
            if (error.type) {
                MEGALogError(@"[Camera Upload] error when to create camera upload node %@", error.nativeError);
                if (self.completion) {
                    self.completion(nil, error.nativeError);
                }
            } else {
                MEGALogDebug(@"[Camera Upload] camera upload node is created");
                MEGANode *node = [MEGASdkManager.sharedMEGASdk nodeForHandle:request.nodeHandle];
                [self saveCameraUploadNode:node];
                if (self.completion) {
                    self.completion(node, nil);
                }
            }
            
            [self finishOperation];
        }];
        
        [MEGASdkManager.sharedMEGASdk createFolderWithName:MEGACameraUploadsNodeName parent:MEGASdkManager.sharedMEGASdk.rootNode
                                                  delegate:delegate];
    }
}

#pragma mark - restore existing camera upload node

- (MEGANode *)restoreExistingCameraUploadNode {
    MEGANode *node = [self savedCameraUploadNodeInLocal];
    if (node == nil) {
        node = [self searchCameraUploadNodeInRoot];
        [self saveCameraUploadNode:node];
    }
    
    return node;
}

- (MEGANode *)savedCameraUploadNodeInLocal {
    unsigned long long cameraUploadHandle = [[[NSUserDefaults standardUserDefaults] objectForKey:CameraUploadsNodeHandleKey] unsignedLongLongValue];
    if (cameraUploadHandle > 0) {
        MEGANode *node = [MEGASdkManager.sharedMEGASdk nodeForHandle:cameraUploadHandle];
        if (node.parentHandle == MEGASdkManager.sharedMEGASdk.rootNode.handle) {
            return node;
        }
    }
    
    return nil;
}

- (MEGANode *)searchCameraUploadNodeInRoot {
    MEGANodeList *nodeList = [MEGASdkManager.sharedMEGASdk childrenForParent:MEGASdkManager.sharedMEGASdk.rootNode];
    NSInteger nodeListSize = [[nodeList size] integerValue];
    
    for (NSInteger i = 0; i < nodeListSize; i++) {
        MEGANode *node = [nodeList nodeAtIndex:i];
        if ([MEGACameraUploadsNodeName isEqualToString:node.name] && node.isFolder) {
            return node;
        }
    }
    
    return nil;
}

- (void)saveCameraUploadNode:(MEGANode *)node {
    if (node == nil) {
        return;
    }
    
    [NSUserDefaults.standardUserDefaults setObject:[NSNumber numberWithUnsignedLongLong:node.handle] forKey:CameraUploadsNodeHandleKey];
}

@end

File: iMEGA/Camera uploads/UploadOperations/CameraUploadNodeLoadOperation.swift
import Foundation

enum NodeLoadError: Error {
    case noRootNode
    case loadCancelled
    case invalidNode
}

final class CameraUploadNodeLoadOperation: MEGAOperation {
    // MARK: - Private properties
    private let completion: NodeLoadCompletion
    private let sdk: MEGASdk
    private let localCachedHandleKey = "CameraUploadsNodeHandle"
    private let autoCreate: Bool
    
    // MARK: - Init
    init(autoCreate: Bool, completion: @escaping NodeLoadCompletion) {
        self.autoCreate = autoCreate
        self.completion = completion
        sdk = MEGASdkManager.sharedMEGASdk()
        super.init()
    }
    
    // MARK: - Life cycle
    override func start() {
        guard !isCancelled else {
            finishOperation(node: nil, error: NodeLoadError.loadCancelled)
            return
        }
        
        startExecuting()
        loadNodeFromRemote()
    }
    
    private func finishOperation(node: MEGANode?, error: Error?) {
        completion(node, error)
        finish()
    }
    
    // MARK: - Load from remote
    private func loadNodeFromRemote() {
        sdk.getCameraUploadsFolder(with: RequestDelegate { [weak self] result in
            switch result {
            case .success(let request):
                self?.validateLoadedHandle(request.nodeHandle)
            case .failure(let error):
                if error.type == .apiENoent {
                    self?.checkLocalCacheForBackwardsCompatibility()
                } else {
                    self?.finishOperation(node: nil, error: error)
                }
            }
        })
    }
    
    // MARK: - Check loaded node handle
    private func validateLoadedHandle(_ handle: NodeHandle) {
        guard let node = handle.validNode(in: sdk) else {
            if autoCreate {
                createCameraUploadNode()
            } else {
                finishOperation(node: nil, error: NodeLoadError.invalidNode)
            }
            return
        }
        
        finishOperation(node: node, error: nil)
    }
    
    // MARK: - Backwards compatibility for local cache
    private func checkLocalCacheForBackwardsCompatibility() {
        guard autoCreate else {
            finishOperation(node: nil, error: NodeLoadError.invalidNode)
            return
        }
        
        guard let cachedHandleNumber = UserDefaults.standard.object(forKey: localCachedHandleKey) as? NSNumber else {
            createCameraUploadNode()
            return
        }
        
        guard let node = cachedHandleNumber.uint64Value.validNode(in: sdk) else {
            createCameraUploadNode()
            return
        }
        
        setCameraUploadTargetFolder(forHandle: node.handle)
    }
    
    private func clearLocalCache() {
        UserDefaults.standard.removeObject(forKey: localCachedHandleKey)
    }
    
    // MARK: - Create Camera Upload Node
    private func createCameraUploadNode() {
        guard let parent = sdk.rootNode else {
            finishOperation(node: nil, error: NodeLoadError.noRootNode)
            return
        }
        
        let name = NSLocalizedString("cameraUploadsLabel", comment: "Name of the auto-generated Camera Upload folder")
        sdk.createFolder(withName: name, parent: parent, delegate: RequestDelegate { [weak self] result in
            switch result {
            case .success(let request):
                self?.setCameraUploadTargetFolder(forHandle: request.nodeHandle)
            case .failure(let error):
                self?.finishOperation(node: nil, error: error)
            }
        })
    }
    
    // MARK: - Set a node handle to be the target folder
    private func setCameraUploadTargetFolder(forHandle handle: NodeHandle) {
        sdk.setCameraUploadsFolderWithHandle(handle, delegate: RequestDelegate { [weak self] result in
            switch result {
            case .success:
                self?.validateLoadedHandle(handle)
                self?.clearLocalCache()
            case .failure(let error):
                self?.finishOperation(node: nil, error: error)
            }
        })
    }
}
